datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -448,7 +474,21 @@ impl ExpressionVisitor for ExprIdentifierVisitor<'_> {
 
         self.id_array[idx] = (self.series_number, desc.clone());
         self.visit_stack.push(VisitRecord::ExprItem(desc.clone()));
-        let data_type = self.data_type.clone();
+
+        let data_type = if let Ok(data_type) = expr.get_type(&self.input_schema) {
+            data_type
+        } else {
+            // expression type could not be resolved in schema, fall back to all schemas
+            let merged_schema =
+                self.all_schemas
+                    .iter()
+                    .fold(DFSchema::empty(), |mut lhs, rhs| {
+                        lhs.merge(rhs);
+                        lhs
+                    });
+            expr.get_type(&merged_schema)?
+        };
```

Comment:
I am honestly unsure if this fall-back logic is necessary at all since the (sub-expression) data types should be resolvable just from the input schema to the respective logical plan node.

However, I did not want to break anything since this fall-back logic was already in place.

All tests pass with and without the fall-back logic.

Reply:
I think all exprs should be resolvable with the unified schemas, as explained above -- but maybe it is a performance optimization 🤔.

Perhaps you could leave a comment explaining that we are not sure if it is necessary

Reply:
This code may be a workaround from some issue we have since fixed 🤔

Reply:
https://github.com/apache/arrow-datafusion/pull/3730 <-- pr to remove the fallback

Reply:
Added comments that explain that the fall-back logic can likely be removed.